devlib/utils/android.py

```diff
@@ -617,15 +617,26 @@ def adb_background_shell(conn, command,
     p = subprocess.Popen(full_command, stdout=stdout, stderr=stderr, stdin=subprocess.PIPE, shell=True)
 
     # Out of band PID lookup, to avoid conflicting needs with stdout redirection
-    find_pid = '{} ps -A -o pid,args | grep {}'.format(conn.busybox, quote(uuid_var))
+    grep_cmd = 'grep {}'.format(quote(uuid_var))
+    find_pid = '{} ps -A -o pid,args | {}'.format(conn.busybox, grep_cmd)
     ps_out = conn.execute(find_pid)
+    # Filter out any commands that contain our grep command.
+    ps_lines = [
+        line.strip() for
+                line in ps_out.splitlines() if not grep_cmd in line
+    ]
     pids = [
-        int(line.strip().split(' ', 1)[0])
-        for line in ps_out.splitlines()
+        int(line.split(' ', 1)[0])
+        for line in ps_lines
     ]
-    # The line we are looking for is the first one, since it was started before
-    # any look up command
-    pid = sorted(pids)[0]
+
+    # Handle the case where the command has already exited otherwise
+    # we should only be left with our desired command
+    if not pids:
+        pid = None
```

Comment:
Should we return `None` or something else here to indicate an invalid PID? Perhaps 0 to be type consistent?

Reply:
In my experience, returning None or a made up value rather than raising does a lot more harm than good, since anything that actually needs the PID will not work with the fake value.

I guess we can end up in this situation if the process does not use stdin/stout/stderr and just exits immediately. On the ssh paramiko connection, PID detection is done by adding a print $$ before exec-ing the command, which is maybe the way to go.

Reply:
> In my experience, returning None or a made up value rather than raising does a lot more harm than good, since anything that actually needs the PID will not work with the fake value.

That's a fair point. My concern over explicitly raising an error is that a usecase I can see for backgrounding a command is that you just want it to execute and don't care if its a short or long running task. In the case that it is short running and we start raising errors this might break existing flows that don't care about the PID. 

> On the ssh paramiko connection, PID detection is done by adding a print $$ before exec-ing the command, which is maybe the way to go.

That seems like a good idea, I'll try switching to that mechanism so that we can try and avoid this issue.

Reply:
> In the case that it is short running and we start raising errors this might break existing flows that don't care about the PID.

I guess exceptions would have to be handled inside devlib mostly, e.g. when getting the PID to kill the command (although I think we would not do anything if the command is already finished). In any case, having a reliable PID would be a better if possible at all. It might complicate things in case of stream redirections though.

Reply:
Agreed I think getting the reliable PID detection is the best way forward here. I've updated the PR with an attempt to align with the mechanism the SSH connection uses. From what I can tell the stream redirection was actually already taking place in this connection as well however the `redirect_streams` function was missing the modification to the returned stream. Please let me know if these changes look reasonable to you.